api/v1beta1/machinedrainrules_types.go

```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	// PodDrainLabel is the label that can be set on Pods in workload clusters to ensure a Pod is not drained.
+	// The only valid value is "skip".
+	// This label takes precedence over MachineDrainRules defined in the management cluster.
+	PodDrainLabel = "cluster.x-k8s.io/drain"
+)
+
+// MachineDrainRuleDrainBehavior defines the drain behavior. Can be either "Drain" or "Skip".
+// +kubebuilder:validation:Enum=Drain;Skip
+type MachineDrainRuleDrainBehavior string
+
+const (
+	// MachineDrainRuleDrainBehaviorDrain means a Pod should be drained.
+	MachineDrainRuleDrainBehaviorDrain MachineDrainRuleDrainBehavior = "Drain"
+
+	// MachineDrainRuleDrainBehaviorSkip means the drain for a Pod should be skipped.
+	MachineDrainRuleDrainBehaviorSkip MachineDrainRuleDrainBehavior = "Skip"
+)
+
+// MachineDrainRuleSpec defines the spec of a MachineDrainRule.
+type MachineDrainRuleSpec struct {
+	// drain configures if and how Pods are drained.
+	// +required
```

Comment:
> The controller tools generator only recently started supporting this version of the required marker

Yup since v0.16.0 from August. But the default behavior is also required. So the generated CRD will look the same even with older controller-gen versions. I think the required marker is mostly for documentation purposes.

> Is the rest of capi using this version or the kubebuilder qualified version?

This repository is using controller-gen v0.16.1. Not sure who else we have to worry about

Reply:
I was mostly looking at other fields and most of them use `+required` instead of `+kubebuilder:validation:Required`.

I don't really have a preference. Do you have one to use one of these or both?

Reply:
I now also took a look at optional. Basically there is one place using the kubebuilder version of the marker. Everywhere else we use either `+required` or `+optional`

Reply:
I don't have a strong preference, just checking in case we want to be consistent.

I have a plan to make a linter for these required/optional tags, I'll maybe bring that up here to stop having to ask the question of which version do we prefer

Reply:
Sounds good!

I think I slightly prefer +required / + optional because they not only work with controller-gen. + I don't really like that all controller-gen markers have a "kubebuilder" prefix :joy: